### PR TITLE
[Feature][AES-1511] Create Product Extract Component for How to Use section

### PR DESCRIPTION
**Pull request by hassanraza0054**
Add Product Extract component for combination with how to use section

1. Product name is a link to its page and gets underlined on hover
2. Product image to the right
3. Header component used for "Works well with"
4. All area starting from the Header down is clickable space for the product link

Add DefinitionListWithProductExtractComponent

1. Showcases current use case for this component
2. Based on requirements, only one product extract at the end of each definition list

**ProductExtract Component Images**

Screen Shot 2020-09-09 at 6 47 57 pm

Screen Shot 2020-09-09 at 6 48 35 pm

Screen Shot 2020-09-09 at 6 49 07 pm

Screen Shot 2020-09-09 at 6 49 35 pm

Screen Shot 2020-09-09 at 6 49 44 pm

Screen Shot 2020-09-09 at 6 49 54 pm

Screen Shot 2020-09-09 at 6 50 10 pm

**DefinitionListWithProductExtract Component Images**

Screen Shot 2020-09-09 at 6 52 00 pm

Screen Shot 2020-09-09 at 6 52 25 pm

Screen Shot 2020-09-09 at 6 52 35 pm

Screen Shot 2020-09-09 at 6 52 49 pm

### CODE REVIEW

**Review by zack76 — COMMENTED** [review]
Looks good, just a few things need to confirm with you

File: src/components/ProductExtract/ProductExtract.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import Image from '~/components/Image';
+import styles from './ProductExtract.module.css';
+import { HEADING } from '~/constants';
+
+import { Heading, Hyperlink } from '~/components';
+
+const ProductExtract = forwardRef(function ProductExtractRef(
+  {
+    className,
+    dataTestRef,
```
[the remark this answers is not in the document]
Comment (hassanraza0054) [reply src/components/ProductExtract/ProductExtract.js:13]: Thanks for that! Fixed

File: src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    url: PropTypes.string,
+  }),
+  works: PropTypes.string,
+  theme: PropTypes.oneOf(['dark', 'light']),
+};
+
+DefinitionListWithProductExtract.defaultProps = {
+  className: undefined,
+  dataTestRef: undefined,
+  items: [{}],
```
[the remark this answers is not in the document]
Comment (hassanraza0054) [reply src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js]: You're right. Empty array works fine too. Fixed

File: src/components/ProductExtract/ProductExtract.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    hasBottomBorder,
+    isVisible,
+    product,
+    theme,
+    itemNum,
```
[the remark this answers is not in the document]
Comment (hassanraza0054) [reply src/components/ProductExtract/ProductExtract.js]: Done

File: src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        works={works}
+      />
+    </Fragment>
+  );
+};
+
+DefinitionListWithProductExtract.propTypes = {
+  className: PropTypes.string,
+  dataTestRef: PropTypes.string.isRequired,
+  productBottomBorder: PropTypes.bool,
```
[the remark this answers is not in the document]
Comment (hassanraza0054) [reply src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js]: Done

File: src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+import DefinitionList from '../DefinitionList/DefinitionList';
+import ProductExtract from '../ProductExtract/ProductExtract';
+
+const DefinitionListWithProductExtract = ({
+  isVisible,
+  items,
+  product,
+  productBottomBorder,
+  itemBottomBorder,
```
[the remark this answers is not in the document]
Comment (hassanraza0054) [reply src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js]: Done

File: src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  works,
+}) => {
+  return (
+    <Fragment>
+      <DefinitionList
+        hasBottomBorder={itemBottomBorder}
+        isVisible={isVisible}
+        items={items}
+      />
+      <ProductExtract
```
Comment (zack76) [inline src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js:25]: wish to double check if there could be only one product be passed in?

**Review by zack76 — APPROVED** [review]

File: src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  itemBottomBorder: false,
+  items: [],
+  productBottomBorder: false,
+  productItemNumber: 1,
+  product: {
+    image: {
+      altText: 'Product Extract',
+      url: '/',
+    },
+  },
```
Comment (kevin-ho87) [inline src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js:68]: Is `name` and `url` defaultProps required in this product object?

File: src/components/ProductExtract/ProductExtract.module.css [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+  .item2 & {
+    transition-delay: calc((2 * 100ms) + 200ms);
+  }
+
+  .item3 & {
+    transition-delay: calc((3 * 100ms) + 200ms);
+  }
+  .item4 & {
+    transition-delay: calc((4 * 100ms) + 200ms);
+  }
```
Comment (kevin-ho87) [inline src/components/ProductExtract/ProductExtract.module.css:47]: Just a thought. Would this be cleaner to have it in a loop?

File: src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js (hunk 7 above)
Comment (hassanraza0054) [reply src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js:68]: I thought url as default was needed but not name for default prop. Could be wrong though

File: src/components/ProductExtract/ProductExtract.module.css (hunk 8 above)
Comment (hassanraza0054) [reply src/components/ProductExtract/ProductExtract.module.css:47]: Probably? But i'm keeping the syntax used consistent with the definition list which also has it written in the same style.

File: src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js (hunk 7 above)
Comment (kevin-ho87) [reply src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js:68]: maybe add `prop: undefined`. Just like the other props. Not sure if it makes any difference :P

**Review by kevin-ho87 — APPROVED** [review]

File: src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  dataTestRef: undefined,
+  isVisible: true,
+  itemBottomBorder: false,
+  items: [],
+  productBottomBorder: false,
+  productItemNumber: 1,
+  product: {
+    image: {
+      altText: 'Product Extract',
+      url: '/',
```
Comment (beau-charman) [inline src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js:66]: I would not pass in a default URL here, either dont render a link, or dont wrap the element in an a tag if there is no URL passed. You can use https://aesop.github.io/aesop-gel/?path=/docs/components-conditionalwrapper--base-component to do so, so in ProductExtract you could do: 

```
<ConditionalWrapper condition={product.url} wrapper={children => (
  <Hyperlink
    className={linkClassSet}
    dataTestRef={dataTestRef}
    title={product.name}
    url={product.url}
  >
    {children}
  </Hyperlink>
))}>
  <>
    <Heading
      className={headingClassSet}
      level={HEADING.LEVEL.FOUR}
      size={HEADING.SIZE.X_X_SMALL}
    >
      <span>{works}</span>
    </Heading>
    <div className={productNameClassSet}>
      <div>
        <span>{product.name}</span>
      </div>
    </div>
    <div className={productImageClassSet}>
      <Image
        altText={product.image.altText}
        small={product.image.small}
      />
    </div>
  <>
</ConditionalWrapper>   
```

File: src/components/ProductExtract/ProductExtract.js [hunk 10]
```diff
@@ -0,0 +1,136 @@
+import React, { forwardRef, Fragment } from 'react';
+import PropTypes from 'prop-types';
+import cx from 'classnames';
+import Image from '~/components/Image';
+import styles from './ProductExtract.module.css';
+import { HEADING } from '~/constants';
+
+import { Heading, Hyperlink } from '~/components';
```
Comment (beau-charman) [inline src/components/ProductExtract/ProductExtract.js:8]: nice 👍  I am glad the named exports are being used from components/index.js. I need to go back through and changed the other components, I just think its nicer than a seperate import for every component

File: src/components/ProductExtract/ProductExtract.js [hunk 11]
```diff
@@ -0,0 +1,136 @@
+import React, { forwardRef, Fragment } from 'react';
```
Comment (beau-charman) [inline src/components/ProductExtract/ProductExtract.js:1]: Could you please order as follows: 

```
import React, { forwardRef, Fragment } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { HEADING } from '~/constants';
import { Heading, Hyperlink, Image } from '~/components';
import styles from './ProductExtract.module.css';
```

File: src/components/ProductExtract/ProductExtract.js [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+  const productImageClassSet = cx(
+    styles.productImage,
+    styles.delayAnimation,
+    {
+      [styles.slideIn]: isVisible,
+    },
+    className,
+  );
+
+  const isEmptyObject = value => {
```
Comment (beau-charman) [inline src/components/ProductExtract/ProductExtract.js:63]: could we either use lodash's version of this, or make it a shared 'util'?

File: src/components/ProductExtract/ProductExtract.js [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+  dataTestRef: undefined,
+  hasBottomBorder: false,
+  isVisible: true,
+  itemNum: 1,
+  product: {
+    image: {
+      altText: 'Product Extract',
+    },
+    name: 'Product Extract',
+    url: '/',
```
Comment (beau-charman) [inline src/components/ProductExtract/ProductExtract.js:130]: Same here with the default url

File: src/components/ProductExtract/ProductExtract.module.css [hunk 14]
```diff
@@ -0,0 +1,127 @@
+@import 'styles/index';
+
+.base {
+  width: 100%;
+  padding: 0;
+  margin: 0;
+  margin-top: 18px;
+  list-style: none;
+  visibility: hidden;
+  @media (--viewport-md) {
```
Comment (beau-charman) [inline src/components/ProductExtract/ProductExtract.module.css:10]: Can we please have new lines before nested rules, like line 20, makes things easier to read :)

**Review by beau-charman — CHANGES_REQUESTED** [review]
Dude, this is great, please review comments

File: src/components/ProductExtract/ProductExtract.module.css (hunk 14 above)
Comment (hassanraza0054) [reply src/components/ProductExtract/ProductExtract.module.css:11]: Done! Sorry about that Beau and thanks a ton for the time and attention to detail.

File: src/components/ProductExtract/ProductExtract.js (hunk 13 above)
Comment (hassanraza0054) [reply src/components/ProductExtract/ProductExtract.js]: Done!

File: src/components/ProductExtract/ProductExtract.js (hunk 12 above)
Comment (hassanraza0054) [reply src/components/ProductExtract/ProductExtract.js]: Switched to Lodash's version and works. Done!

File: src/components/ProductExtract/ProductExtract.js (hunk 11 above)
Comment (hassanraza0054) [reply src/components/ProductExtract/ProductExtract.js]: Ordered accordingly. Added the Lodash import before **cx** using the `UserSession.js` component as reference.

image

File: src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js (hunk 9 above)
Comment (hassanraza0054) [reply src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js]: Done! Thanks Beau

**Review by beau-charman — APPROVED** [review]
🙌

File: src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js (hunk 7 above)
Comment (hassanraza0054) [reply src/components/DefinitionListWithProductExtract/DefinitionListWithProductExtract.js]: All good! Stuck with removing url for this one and keeping the rest as is. Thanks for the input Kev!

**Pull request merged**